primitives/transaction-storage-proof/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storge proof primitives. Constains types and basic code to extract storage
+//! proofs for indexed transactions.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::{result::Result, prelude::*};
+
+use codec::{Encode, Decode};
+use sp_inherents::{Error, InherentIdentifier, InherentData, IsFatalError};
+
+/// The identifier for the proof inherent.
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"tx_proof";
+/// Storage period for data.
+pub const DEFAULT_STORAGE_PERIOD: u32 = 100800;
+/// Proof trie value size.
+pub const CHUNK_SIZE: usize = 256;
+
+/// Errors that can occur while checking the storage proof.
+#[derive(Encode, sp_runtime::RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Decode))]
+pub enum InherentError {
+	InvalidProof,
+	TrieError
+}
+
+impl IsFatalError for InherentError {
+	fn is_fatal_error(&self) -> bool {
+		true
+	}
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Debug)]
+pub struct TransactionStorageProof {
+	/// Data chunk that is proved to exist.
+	pub chunk: Vec<u8>,
+	/// Trie nodes that compose the proof.
+	pub proof: Vec<Vec<u8>>,
+}
+
+/// Auxiliary trait to extract storage proof.
+pub trait TransactionStorageProofInherentData {
+	/// Get the proof.
+	fn storage_proof(&self) -> Result<Option<TransactionStorageProof>, Error>;
+}
+
+impl TransactionStorageProofInherentData for InherentData {
+	fn storage_proof(&self) -> Result<Option<TransactionStorageProof>, Error> {
+		Ok(self.get_data(&INHERENT_IDENTIFIER)?)
+	}
+}
+
+/// Provider for inherent data.
+#[cfg(feature = "std")]
+pub struct InherentDataProvider<F> {
+	inner: F,
+}
+
+#[cfg(feature = "std")]
+impl<F> InherentDataProvider<F> {
+	pub fn new(proof_oracle: F) -> Self {
+		InherentDataProvider { inner: proof_oracle }
+	}
+}
+
+#[cfg(feature = "std")]
+impl<F> sp_inherents::ProvideInherentData for InherentDataProvider<F>
+where F: Fn() -> Result<Option<TransactionStorageProof>, Error>
+{
+	fn inherent_identifier(&self) -> &'static InherentIdentifier {
+		&INHERENT_IDENTIFIER
+	}
+
+	fn provide_inherent_data(&self, inherent_data: &mut InherentData) -> Result<(), Error> {
+		let proof = (self.inner)()?;
+		if let Some(proof) = proof {
+			inherent_data.put_data(INHERENT_IDENTIFIER, &proof)
+		} else {
+			Ok(())
+		}
+	}
+
+	fn error_to_string(&self, _error: &[u8]) -> Option<String> {
+		Some(format!("no further information"))
+	}
+}
+
+/// An utility function to extract chunk index from the source of randomness.
+pub fn random_chunk(random_hash: &[u8], total_chunks: u32) -> u32 {
+	let mut buf = [0u8; 8];
+	buf.copy_from_slice(&random_hash[0..8]);
+	let random_u64 = u64::from_be_bytes(buf);
+	(random_u64 % total_chunks as u64) as u32
```

Comment:
Should we configure a number of checked chunk (not only one) depending on the total _chuncks?

Reply:
1 should be good enough to discourage validators from messing around with data and potentially losing bock rewards. We can increase it later if required.